From 531bdbf18cc5aa001b72281dc964f1aea3d78cb7 Mon Sep 17 00:00:00 2001
From: Haris Okanovic <haris.okanovic@ni.com>
Date: Tue, 29 Sep 2015 10:42:16 -0500
Subject: [PATCH 1/2] rtctl: Read task IDs from procfs instead of `ps`

Read task IDs from procfs instead of running `ps -eo pid,cmd`
which is not implemented by BusyBox.

Upstream-Status: Pending

Signed-off-by: Haris Okanovic <haris.okanovic@ni.com>
---
 rtctl | 28 +++++++++++++++++++++++-----
 1 file changed, 23 insertions(+), 5 deletions(-)

diff --git a/rtctl b/rtctl
index 9ccd504..be09b00 100755
--- a/rtctl
+++ b/rtctl
@@ -4,7 +4,7 @@ usage ()
 {
     echo "Usage: $0 [--file <filename>] set <groupname> <prio> <fifo|batch|rr|other>"
     echo "       $0 [--file <filename>] reset [<groupname>]"
-    echo "       $0 [--file <filename>] show <groupname> [<ps_fmt>]"
+    echo "       $0 [--file <filename>] show <groupname>"
     exit 1
 }
 
@@ -43,6 +43,24 @@ fi
 CMD=$1
 shift
 
+PROCFS_PATH="/proc"
+
+#
+# Prints each running PID and command line
+# Equivalent to `ps -eo pid,cmd`, but filters out user space tasks
+#
+get_kernel_tasks ()
+{
+    ls "$PROCFS_PATH" | egrep '^[0-9]+$' |
+    while read procId; do
+        procPth="$PROCFS_PATH/$procId"
+        if [ ! -s "$procPth/exe" -a ! -s "$procPth/cmdline" -a -s "$procPth/comm" ]; then
+            read comm <"$procPth/comm" || comm=""
+            echo "$procId [$comm]"
+        fi
+    done;
+}
+
 GROUPNAME=""
 
 #
@@ -51,7 +69,7 @@ GROUPNAME=""
 #
 group_pids ()
 {
-  ps -eo pid,cmd | awk '
+  get_kernel_tasks | awk '
     /^[a-zA-Z_0-9-]+:[*orbf]:[0-9]+:.+$/ {
       split($0, parts, ":") 
       if (parts[1] == groupname) {
@@ -76,7 +94,7 @@ group_pids ()
 
 set_group_defaults ()
 {
-  ps -eo pid,cmd | awk '
+  get_kernel_tasks | awk '
     /^[a-zA-Z_0-9-]+:[*orbf]:[0-9]+:.+$/ {
       split($0, conf, ":")
       if (groupname == "" || conf[1] == groupname) {
@@ -110,12 +128,12 @@ set_group_defaults ()
 
 
 #
-# show "ps" output for group of threads
+# show group of threads
 #
 do_show ()
 {
     ps_fmt=$1
-    ( group_pids; ps -eo $ps_fmt ) | awk '
+    ( group_pids; get_kernel_tasks ) | awk '
     /^[0-9]+$/ {
       pids[$1] = 1
     }
-- 
2.12.2

